### PR TITLE
[cudax] Initial `cudax::coop::reduce` prototype

### LINKED ISSUE #9156: [FEA]: Initial `cudax::coop::reduce` prototype
Create a simplified implementation that takes `cudax::this_(thread|warp|block)` as the parameter.

### PR DESCRIPTION
**Pull request by davebayer**
This PR implements `cudax::coop::reduce` function prototype for `cudax::this_(thread|warp|block)`. The function dispatches to `cub::` equivalents for now and allocates all of the shared memory itself.

### CODE REVIEW

File: cub/cub/warp/warp_reduce.cuh [hunk 1]
```diff
@@ -504,7 +504,7 @@ public:
   [[nodiscard]] _CCCL_DEVICE _CCCL_FORCEINLINE T Reduce(const InputType& input, ReductionOp reduction_op)
   {
     auto thread_reduction = cub::ThreadReduce(input, reduction_op);
-    return WarpReduce<T, LogicalWarpThreads>::Reduce(thread_reduction, LogicalWarpThreads, reduction_op);
+    return WarpReduce<T, LogicalWarpThreads>::Reduce(thread_reduction, reduction_op);
```
Comment (miscco) [inline cub/cub/warp/warp_reduce.cuh:507]: Please move into separate PR and rebase once merged

File: cudax/cmake/cudaxBuildCompilerTargets.cmake [hunk 2]
```diff
@@ -35,7 +35,7 @@ if ("Clang" STREQUAL "${CMAKE_CXX_COMPILER_ID}")
   append_option_if_available("-Wno-deprecated-copy" cxx_compile_options)
 endif()
 
-list(APPEND cxx_compile_definitions CCCL_ENABLE_ASSERTIONS)
+list(APPEND cxx_compile_definitions CCCL_ENABLE_ASSERTIONS _CUDAX_GROUP)
```
Comment (miscco) [inline cudax/cmake/cudaxBuildCompilerTargets.cmake]: I would love a tad more descriptive macro like `_CUDAX_ENABLE_GROUP`

File: cudax/include/cuda/experimental/__coop/reduce.cuh [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+template <class _Hierarchy, class _Tp, ::cuda::std::size_t _Np, class _RedFn>
+[[nodiscard]] _CCCL_DEVICE_API ::cuda::std::optional<_Tp>
+__reduce_impl(this_warp<_Hierarchy> __group, _Tp (&__thread_data)[_Np], _RedFn __red_fn)
+{
+  using _WarpReduce = ::cub::WarpReduce<_Tp>;
+  __shared__ typename _WarpReduce::TempStorage __scratch;
+
+  const auto __result = _WarpReduce{__scratch}.Reduce(__thread_data, __red_fn);
+  return (gpu_thread.is_root_rank(__group)) ? ::cuda::std::optional{__result} : ::cuda::std::nullopt;
```
Comment (miscco) [inline cudax/include/cuda/experimental/__coop/reduce.cuh:55]: I am confused by this line. Is this because the value is only valid in the leader thread? Should we broadcast it rather than diverging further?

Otherwise, why do we even compute it if its not desired

File: cudax/test/coop/reduce/this_thread.cu [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+  template <class Config, int NumItems, class T, class RedOp>
+  __device__ void operator()(
+    Config config,
+    cuda::std::integral_constant<int, NumItems>,
+    const T* __restrict__ d_in,
+    T* __restrict__ d_out,
+    RedOp red_op)
+  {
+    T thread_data[NumItems];
+#pragma unroll
```
Comment (miscco) [inline cudax/test/coop/reduce/this_thread.cu]: Shouldnt this just use the CCCL macros for that?

File: cudax/test/coop/reduce/this_thread.cu [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+};
+
+template <>
+struct operator_to_std<cuda::maximum<>>
+{
+  using type = cuda::maximum<>;
+};
+
+template <class CudaT>
+using operator_to_std_t = typename operator_to_std<CudaT>::type;
```
Comment (miscco) [inline cudax/test/coop/reduce/this_thread.cu]: Is this used anywhere?

File: cub/cub/warp/warp_reduce.cuh (hunk 1 above)
Comment (davebayer) [reply cub/cub/warp/warp_reduce.cuh:507]: Implemented by #9153

File: cudax/cmake/cudaxBuildCompilerTargets.cmake (hunk 2 above)
Comment (davebayer) [reply cudax/cmake/cudaxBuildCompilerTargets.cmake]: Yeah, I thought the same thing

Comment (davebayer) [reply cudax/cmake/cudaxBuildCompilerTargets.cmake]: I extracted this to a separate PR #9180

File: cudax/include/cuda/experimental/__coop/reduce.cuh (hunk 3 above)
Comment (davebayer) [reply cudax/include/cuda/experimental/__coop/reduce.cuh:55]: This is because only the root rank has the correct value. CUB does the same thing, but returns garbage for non-root ranks. I think returning optional is the better way.

We've already discussed this before and we agreed that we would start like this and add an API that would also broadcast the result in the future.

File: cudax/test/coop/reduce/this_thread.cu (hunk 4 above)
Comment (davebayer) [reply cudax/test/coop/reduce/this_thread.cu]: I'll just get rid of the pragma completely, it's irrelevant for the test

File: cudax/test/coop/reduce/this_thread.cu [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+constexpr int max_size  = 16;
+constexpr int num_seeds = 10;
+
+/***********************************************************************************************************************
+ * Test cases
+ **********************************************************************************************************************/
+
+_CCCL_DIAG_SUPPRESS_MSVC(4244) // warning C4244: '=': conversion from 'int' to '_Tp', possible loss of data
+
+C2H_TEST("reduce/this_thread Integral Type Tests", "[reduce][this_thread]", integral_type_list, operator_integral_list)
```
Comment (pciolkosz) [inline cudax/test/coop/reduce/this_thread.cu:165]: The tests look very similar across the hierarchy levels, could it be just a template? The test body is mostly the same, the dispatch is almost the same

**Review by pciolkosz — APPROVED** [review]

File: cudax/test/coop/reduce/this_thread.cu (hunk 6 above)
Comment (davebayer) [reply cudax/test/coop/reduce/this_thread.cu:165]: It's simplified a bit for now, I will need to extend the tests in the future to be more specific for each level, so I'd rather duplicate the code for now
